\section{Background}
\label{sec:back}

Genome sequencers read DNA\nomenclature{DNA}{Deoxyribonucleic acid} strands in segments and convert each segment into a packet of data known as a \textit{read}\nomenclature{Read}{Data generated from sequencing a segment of DNA}. DNA sequencing devices from \href{https://nanoporetech.com/}{Oxford Nanopore Technologies} (ONT)\nomenclature{ONT}{Oxford Nanopore Technologies} are no different and record disturbances in ionic current as DNA molecules are passed through a biological nanopore \cite{fast5}. These measurements can be translated to determine the sequence of each DNA molecule analysed.

The time series current signal data is written in a format called FAST5, which stores the raw data for each nanopore sequencing read. FAST5 is a Hierarchical Data Format 5 (HDF5)\nomenclature{HDF5}{Hierarchical Data Format 5} file \cite{hdf5:specs} with a specific schema defined by ONT. HDF5 is a complex file format for storing and managing high volume data. It works well with the time series current signal of a nanopore read, since it uses B-trees to index table objects.

However, the official library used to access the HDF5 file format is not scalable with threads \cite{hdf5:nomulti}. This means that tasks performed with the FAST5 file format suffer from an inefficient utilisation of parallel resources. Therefore, the process of basecalling FAST5 data into DNA sequence reads and other common analyses that utilise signal-level data (such as DNA methylation calling with Nanopolish \cite{nanopolish} or f5c \cite{f5c}) slow down the typical sequencing pipeline.
